python/tvm/script/context_maintainer.py

```diff
@@ -16,59 +16,145 @@
+from tvm.runtime import Object
+from .node import BufferSlice
+
+
+class BlockInfo:
+    """Information for block and block_realize signature"""
+
+    alloc_buffers: List[Buffer]
+    match_buffers: List[MatchBufferRegion]
+    iter_bindings: Mapping[Var, PrimExpr]
+    reads: Optional[List[BufferSlice]]
+    writes: Optional[List[BufferSlice]]
+    annotations: Optional[Mapping[str, Object]]
+    predicate: Optional[PrimExpr]
+    init: Optional[Stmt]
+
+    def __init__(self):
+        self.alloc_buffers = []
+        self.match_buffers = []
+        self.iter_bindings = {}
+        self.reads = None
+        self.writes = None
+        self.annotations = None
+        self.predicate = None
+        self.init = None
 
 
 class ContextMaintainer:
     """Maintain all the necessary context info"""
 
-    def __init__(self, parser):
+    # scope context
+    # ast_node inside a scope
+    node_stack: List[List[synr.ast.Node]]
+    # loop stacks inside a block
+    block_info_stack: List[BlockInfo]
+    # loop stacks inside a block
+    loop_stack: List[List[Var]]
+    symbols: List[Dict[str, Union[Var, Buffer]]]
+
+    # function context
+    func_params: List[Var]
+    func_buffer_map: Mapping[Var, Buffer]
+    func_dict_attr: Mapping[str, Object]
+    func_var_env_dict: Mapping[Var, str]
+
+    # parser and analyzer
+    _report_error: Callable[[str, Union[Span, synr.ast.Span]], None]
+    analyzer: tvm.arith.Analyzer
+
+    def __init__(self, _report_error: Callable[[str, Union[Span, synr.ast.Span]], None]):
         # scope context
         self.node_stack = []  # AST nodes of scopes
+        self.block_info_stack = []  # Block info of scopes
+        self.loop_stack = []  # stack of loop vars
         self.symbols = []  # symbols of scopes
         # function context
         self.func_params = []  # parameter list of function
         self.func_buffer_map = {}  # buffer_map of function
         self.func_dict_attr = {}  # func_attr of function
         self.func_var_env_dict = {}  # map from var to env_name
-        # parser
-        self.parser = parser
+        # parser and analyzer
+        self._report_error = _report_error
+        self.analyzer = tvm.arith.Analyzer()
 
-    def pop_scope(self):
-        """Pop the inner most scope"""
-        self.symbols.pop()
-        self.node_stack.pop()
+    def enter_scope(self, nodes: Optional[List[synr.ast.Node]] = None):
+        """Creating a new scope
 
-    def new_scope(self, nodes=None):
-        """Creating a new scope"""
+        Parameters
+        ----------
+        nodes : Optional[List[synr.ast.Node]]
+            The synr AST nodes in new scope
+        """
         if nodes is None:
             nodes = []
         self.node_stack.append(list(reversed(nodes)))
         self.symbols.append(dict())
 
-    def update_symbol(self, name, symbol):
+    def enter_block_scope(self, nodes: Optional[List[synr.ast.Node]] = None):
```

Comment:
Document the difference between a regular and block scope from the user perspective. When should a should `enter_scope` be used and when should `enter_block_scope` be used.